pkg/smokescreen/smokescreen_test.go

```diff
@@ -1019,6 +1021,121 @@ func TestRejectResponseHandler(t *testing.T) {
 	})
 }
 
+func TestVerifyRequestHandler(t *testing.T) {
+	r := require.New(t)
+	testHeader := "X-Verify-Request-Header"
+	h := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("OK"))
+	})
+	v := func(r *http.Request) error {
+		header := r.Header.Get(testHeader)
+		if header == "" {
+			return errors.New("header doesn't exist")
+		}
+		if header != "valid" {
+			return errors.New("invalid header")
+		}
+		return nil
+	}
+
+	t.Run("Testing that verify request handler works for HTTPS", func(t *testing.T) {
+		testCases := []struct {
+			header        http.Header
+			expectedError bool
+		}{
+			{
+				header:        http.Header{testHeader: []string{"valid"}},
+				expectedError: false,
+			},
+			{
+				header:        http.Header{testHeader: []string{"invalid"}},
+				expectedError: true,
+			},
+		}
+		cfg, err := testConfig("test-local-srv")
+		r.NoError(err)
+		err = cfg.SetAllowAddresses([]string{"127.0.0.1"})
+		r.NoError(err)
+		cfg.VerifyRequestHandler = v
+
+		l, err := net.Listen("tcp", "localhost:0")
+		r.NoError(err)
+		cfg.Listener = l
+
+		proxy := proxyServer(cfg)
+		remote := httptest.NewTLSServer(h)
+		defer proxy.Close()
+		for _, testCase := range testCases {
+
+			client, err := proxyClient(proxy.URL, testCase.header)
+			r.NoError(err)
+
+			req, err := http.NewRequest("GET", remote.URL, nil)
+			r.NoError(err)
+			resp, err := client.Do(req)
+			if testCase.expectedError {
+				r.Nil(resp)
+				r.Contains(err.Error(), "Request rejected by proxy")
+			} else {
+				r.NoError(err)
+				r.Equal(200, resp.StatusCode)
+			}
+		}
+	})
+
+	t.Run("test that it works for HTTP (invalid)", func(t *testing.T) {
+		testCases := []struct {
+			header        string
+			expectedError bool
+		}{
+			{
+				header:        "valid",
+				expectedError: false,
+			},
+			{
+				header:        "invalid",
+				expectedError: true,
+			},
+		}
+		cfg, err := testConfig("test-local-srv")
+		r.NoError(err)
+		err = cfg.SetAllowAddresses([]string{"127.0.0.1"})
+		r.NoError(err)
+		cfg.VerifyRequestHandler = v
+
+		l, err := net.Listen("tcp", "localhost:0")
+		r.NoError(err)
+		cfg.Listener = l
+
+		remote := httptest.NewServer(h)
+
+		proxySrv := proxyServer(cfg)
+		r.NoError(err)
+		defer proxySrv.Close()
+
+		// Create a http.Client that uses our proxy
+		client, err := proxyClient(proxySrv.URL, nil)
+		r.NoError(err)
+
+		for _, testCase := range testCases {
+			// Send a request that should be blocked
+			req, err := http.NewRequest("GET", remote.URL, nil)
+			r.NoError(err)
+			req.Header.Set(testHeader, testCase.header)
```

Comment:
We're passing an extra header to conduct some verification here. We should remove it from the request before proxying it further as we don't want to leak internal details or pass unnecessary information.

Could you update the test to demonstrate that it's possible to remove the header from the original request before proxying it?

Reply:
Removing the header in the handler now.

Reply:
https://github.com/stripe/smokescreen/pull/180/commits/382d64a93b264ff460417837bf4fe4efe130a281